### PR TITLE
Implemented verify_header method

### LINKED ISSUE #9: Implement header verification
Static header checks. See stubbed function `verify_header`.

### PR DESCRIPTION
**Pull request by voith**
### What was wrong?

closes #9

### How was it fixed?
- Renamed `Header` attributes to match the names mentioned in the yellow paper.
- Implemented `verify_header` method as specified in the YP as shown below:

Screenshot 2021-06-11 at 2 01 25 PM

### TODO
- [x] Cleanup code, Add annotations and add docstrings to methods.
- [x] Verify that spec is rendered as expected.
- [x] Explain to the team why `n <= 2** 256 // Hd` and m = Hm check has not been implemented. 
- [x] Add tests

### CODE REVIEW

**Comment by lightclient** [comment]
Looking good so far! I converted this to a draft, feel free to convert to an open PR when you're ready. Thanks.

File: src/eth1spec/spec.py [hunk 1]
```diff
@@ -321,6 +329,56 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+    return True
+
+def calculate_block_difficulty(header: Header, parent_header: Header) -> bool:
+    if parent_header.number == 0:
+        return 131072
+    elif header.timestamp < parent_header.timestamp:
+        return parent_header.difficulty + parent_header.difficulty // 2048
+    else:
+        return map(131072, parent_header.difficulty - parent_header.difficulty // 2048)
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]:
```suggestion
        return max(131072, parent_header.difficulty - parent_header.difficulty // 2048)
```

Comment (voith) [reply src/eth1spec/spec.py]: Thanks for catching this. I had this fixed on my local machine.

**Comment by lightclient** [comment]
Hey @voith just checking if you're still working on this?

**Comment by voith** [comment]
Hey @lightclient. I fell sick. Sorry for the silence. I'm working on this right now. I will finish this by EOD.

**Comment by lightclient** [comment]
@voith no worries! Hope you're okay!

File: src/eth1spec/spec.py [hunk 2]
```diff
@@ -35,6 +35,8 @@
 from .evm.interpreter import process_call
 
 BLOCK_REWARD = 5 * 10 ** 18
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]: *nitpick*

Not part of this PR but, some alternate ideas that come to mind for this variable.

```python
BLOCK_REWARD = 5000000.....00000  # 5 * 10**18
# or
BLOCK_REWARD = 5 * (10 ** 18)
# or
ONE_ETHER = 1**18
BLOCK_REWARD = 5 * ONE_ETHER
```

Comment (voith) [reply src/eth1spec/spec.py]: I like `BLOCK_REWARD = 5 * ONE_ETHER`, it's more readable

File: src/eth1spec/spec.py [hunk 3]
```diff
@@ -75,21 +78,30 @@ def state_transition(chain: BlockChain, block: Block) -> None:
...
     Returns
     -------
     verified : `bool`
         True if the header is correct, False otherwise.
     """
-    raise NotImplementedError()  # TODO
+    assert header.difficulty == calculate_block_difficulty(
+        header, parent_header
+    )
+    assert validate_gas_limit(header.gas_limit, parent_header.gas_limit)
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]: > This comment extends beyond this PR and thus shouldn't block the merge of this PR.

Whatever is chosen, we should be really consistent about this.  In general, I've found validation functions that return booleans to be foot-guns, but that stems from working for a long time under the convention that validation functions should raise exceptions or return `None`.  This is of course heavily influenced by writing "library" code, which this is not....

boolean return pattern is pretty easy to understand, and this model of many lines of `assert, assert, assert` is visually consistent.  I think it would be worth documenting some of the "style/convention" choices so that there is consistency in how the various pieces of this code get written.

File: src/eth1spec/spec.py [hunk 4]
```diff
@@ -330,6 +342,126 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    Returns
+    -------
+    Header : `eth1spec.eth_types.Header`
+        Block header found by its hash.
+    """
+    for block in chain.blocks:
+        if header_hash(block.header) == hash:
+            return block.header
+    else:
+        raise ValueError("Could not find Header with hash {!r}".format(hash))
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]: >  Again, not really something specifically for this PR but more a style guide consistency comment for this whole repository.

Curious of the choice of `.format(...)` over f-strings.

Comment (voith) [reply src/eth1spec/spec.py]: flake started complaining when I used f-strings. I can address this in a separate PR.

File: src/eth1spec/spec.py [hunk 5]
```diff
@@ -330,6 +342,126 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    verified : `bool`
+        True if header gas limit constraints are satisfied, False otherwise
+    """
+    if gas_limit >= parent_gas_limit + (
+        parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    ):
+        return False
+    if gas_limit <= parent_gas_limit - (
+        parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    ):
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]: maybe more readable as:

```python
max_adjustment_delta = parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
if gas_limit >= parent_gas_limit + max_adjustment_delta:
    return False
if gas_limit <= parent_gas_limit - max_adjustment_delta:
    return False
```

File: src/eth1spec/spec.py [hunk 6]
```diff
@@ -330,6 +342,126 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    difficulty: Uint
+        computed difficulty for a block
+    """
+    if parent_header.number == 0:
+        return Uint(131072)
+    elif header.timestamp < parent_header.timestamp:
+        return parent_header.difficulty + parent_header.difficulty // Uint(
+            2048
+        )
+    else:
```
Comment (pipermerriam) [inline src/eth1spec/spec.py]: Another style choice that extends beyond this repository.  This else block is an *implicit* conditional for `elif header.timestamp > parent_header.timestamp`.  But the condition isn't actually written which makes understanding the logic just a tiny bit harder.  Options for improving this:

```python
# use a comment
if thing == 1:
    ...
elif thing > 1:
    ...
else:  # thing < 1
    ...
    
# explicit invariant
if thing == 1:
    ...
elif thing > 1:
    ...
elif thing < 1:
    ...
else:  # unreachable
    raise Exception("Unreachable")
```

File: src/eth1spec/spec.py (hunk 3 above)
Comment (SamWilsn) [reply src/eth1spec/spec.py]: 110% agree we need to document our coding conventions. Opened #42 to track.

File: src/eth1spec/spec.py (hunk 4 above)
Comment (SamWilsn) [reply src/eth1spec/spec.py]: Was it complaining inside of `tox`, or are you running it outside the build environment?

**Comment by voith** [comment]
@lightclient @SamWilsn
I haven't implemented `n <= 2** 256 // Hd and m = Hm` check as I will need to implement [EthHash](https://eth.wiki/en/concepts/ethash/ethash) to do so. I think that's beyond the scope of this PR and I can implement it in a follow-up PR. What do you think?

File: src/eth1spec/spec.py (hunk 6 above)
Comment (SamWilsn) [reply src/eth1spec/spec.py]: I don't think I like either of these styles. Personally, I think most developers understand if/elif/else statements fairly well.

If the conditions ever change (say in an EIP), the comment could get out of sync with the code, and I don't think the tooling we have can actually assert that the `else: # unreachable` is actually unreachable.

**Comment by SamWilsn** [comment]
> @lightclient @SamWilsn
> I haven't implemented `n <= 2** 256 // Hd and m = Hm` check as I will need to implement [EthHash](https://eth.wiki/en/concepts/ethash/ethash) to do so. I think that's beyond the scope of this PR and I can implement it in a follow-up PR. What do you think?

I think that's fine. Maybe add a `TODO` comment where the check should eventually go (if you haven't already.)

File: src/eth1spec/spec.py (hunk 2 above)
Comment (lightclient) [reply src/eth1spec/spec.py]: I do as well.

File: src/eth1spec/spec.py (hunk 4 above)
Comment (lightclient) [reply src/eth1spec/spec.py]: Huh, they seemed to be working here: https://github.com/quilt/eth1.0-specs/blob/pyspec/tests/test_trie.py#L20

File: src/eth1spec/spec.py (hunk 6 above)
Comment (pipermerriam) [reply src/eth1spec/spec.py]: I don't feel strongly about this.  It's simply a convention that we adopted on our team as a standard convention due to a few times that we've run into bugs due to implicit else clauses that either never get hit *or* that are hard to reason about.  This isn't really either of those.

Regarding tooling:  I don't think this actually applies but here's the best I'm aware of for exhaustiveness checking in python: https://hakibenita.com/python-mypy-exhaustive-checking

File: src/eth1spec/spec.py (hunk 3 above)
Comment (pipermerriam) [reply src/eth1spec/spec.py]: The maximum amount of agreement is 100%.  I call malarkey, shenanigans, or potentially some form of witchcraft.

File: src/eth1spec/spec.py (hunk 6 above)
Comment (SamWilsn) [reply src/eth1spec/spec.py]: Oh that's clever! I'll play around with that.

File: src/eth1spec/spec.py (hunk 4 above)
Comment (voith) [reply src/eth1spec/spec.py]: It's `mypy` complaining for some reason, not flake8. I get the following when I run `tox -e py3`:
```
py3 run-test: commands[3] | mypy src tests setup.py --exclude 'tests/fixtures/*'
src/eth1spec/spec.py:403: error: On Python 3 '{}'.format(b'abc') produces "b'abc'", not 'abc'; use '{!r}'.format(b'abc') if this is desired behavior
```
-------
Upon investigating I realized that I need convert the hash value(bytes) to hex for the sake of readability.

**Comment by voith** [comment]
@SamWilsn @lightclient Ready for review.

**Comment by voith** [comment]
@SamWilsn @lightclient This PR has some merge conflicts. But it seems like the two of you are busy addressing Pipers feedback. I will wait till you finish. Let me know when you're ready to review. I will resolve the conflicts then.

**Comment by lightclient** [comment]
@voith if you want to go ahead now, thanks!

**Comment by voith** [comment]
@lightclient resolved merged conflicts, ready for review.

File: src/eth1spec/spec.py [hunk 7]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
     )
 
 
+def header_hash(header: Header) -> Hash32:
```
Comment (lightclient) [inline src/eth1spec/spec.py]: Since we're going with active naming, we should probably make this something like `calculate_header_hash` or `computer_header_hash`.

File: src/eth1spec/spec.py [hunk 8]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    """
+    Computes the hash of a block Header.
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
    Computes the hash of a block header.
```
I don't think this needs to be capitalized?

File: src/eth1spec/spec.py [hunk 9]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+    Parameters
+    ----------
+    header : `eth1spec.eth_types.Header`
+        Header of interest.
+
+    Returns
+    -------
+    hash : `eth1spec.eth_types.Hash32`
+        Hash of the Header.
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        Hash of the header.
```
same here?

File: src/eth1spec/spec.py [hunk 10]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+
+def get_block_header_by_hash(hash: Hash32, chain: BlockChain) -> Header:
+    """
+    Fetches the block header by searching its hash.
+
+    Parameters
+    ----------
+    hash : `eth1spec.eth_types.Hash32`
+        hash of the header of interest.
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        Hash of the header of interest.
```

File: src/eth1spec/spec.py [hunk 11]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+                header.mix_digest,
+                header.nonce,
+            )
+        )
+    )
+
+
+def get_block_header_by_hash(hash: Hash32, chain: BlockChain) -> Header:
+    """
+    Fetches the block header by searching its hash.
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
    Fetches the block header with the corresponding hash.
```

File: src/eth1spec/spec.py [hunk 12]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    Returns
+    -------
+    Header : `eth1spec.eth_types.Header`
+        Block header found by its hash.
+    """
+    for block in chain.blocks:
+        if header_hash(block.header) == hash:
+            return block.header
+    else:
+        raise ValueError(f"Could not find Header with hash={hash.hex()}")
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        raise ValueError(f"Could not find header with hash={hash.hex()}")
```

File: src/eth1spec/spec.py [hunk 13]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    ----------
+    gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the header of interest
+
+    parent_gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the porent header of interest
+
+    Returns
+    -------
+    verified : `bool`
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
    valid : `bool`
```

File: src/eth1spec/spec.py [hunk 14]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+
+def validate_gas_limit(gas_limit: Uint, parent_gas_limit: Uint) -> bool:
+    """
+    Validates the gas limit for a block
+
+    Parameters
+    ----------
+    gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the header of interest
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        Gas limit to validate.
```

File: src/eth1spec/spec.py [hunk 15]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+    parent_gas_limit : `eth1spec.eth_types.Uint`
+        gas limit of the porent header of interest
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        Gas limit of the parent block.
```

File: src/eth1spec/spec.py [hunk 16]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+
+    Returns
+    -------
+    verified : `bool`
+        True if header gas limit constraints are satisfied, False otherwise
```
Comment (lightclient) [inline src/eth1spec/spec.py]:
```suggestion
        True if gas limit constraints are satisfied, False otherwise
```

File: src/eth1spec/spec.py [hunk 17]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+        return False
+    if gas_limit <= parent_gas_limit - max_adjustment_delta:
+        return False
+    if gas_limit < GAS_LIMIT_MINIMUM:
+        return False
+
+    return True
+
+
+def calculate_block_difficulty(header: Header, parent_header: Header) -> Uint:
```
Comment (lightclient) [inline src/eth1spec/spec.py]: Seems like you only need the parent's number, timestamp, difficult, and the current timestamp -- is there a reason to not just accept those parameters instead both headers?

File: tests/test_validate_header.py [hunk 18]
```diff
@@ -0,0 +1,174 @@
+import pytest
+
+from eth1spec.base_types import U256, Bytes, Bytes8, Bytes32
+from eth1spec.eth_types import Address, Bloom, Hash32, Header, Root, Uint
+from eth1spec.spec import verify_header
+from tests.helpers import hex2hash
+
+
+def create_fake_parent_header(
```
Comment (lightclient) [inline tests/test_validate_header.py]: Generally I think we should avoid writing our own tests at all costs. The `ethereum/tests` repo provides some blockchain tests that we could use in place: https://github.com/ethereum/tests/tree/develop/LegacyTests/Constantinople/BlockchainTests

Comment (lightclient) [reply tests/test_validate_header.py]: It would be good to replace these tests with a runner for some of those tests.

**Review by lightclient — CHANGES_REQUESTED** [review]
This is looking good @voith, thank you! My main concern is writing custom unit tests. I feel like we should avoid writing our own tests at all costs, because ideally spec tests should be trivially run by other clients. Since `ethereum/tests` already has an accepted cross client format we should follow their lead.

File: src/eth1spec/spec.py (hunk 7 above)
Comment (Bhargavasomu) [reply src/eth1spec/spec.py]: I like `compute_header_hash`

File: src/eth1spec/spec.py [hunk 19]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    parent_header: `eth1spec.eth_types.Header`
+        parent header of header of interest
+    Returns
+    ------
+    difficulty: Uint
+        Computed difficulty for a block.
+    """
+    if parent_header.number == 0:
+        return Uint(131072)
+    elif header.timestamp < parent_header.timestamp:
```
Comment (Bhargavasomu) [inline src/eth1spec/spec.py]: As per the frontier yellowpaper, shouldn't this be as follows?
```
elif header.timestamp < parent_header.timestamp + 13:
    ...
```
Please correct me if wrong.
\cc @lightclient @SamWilsn

File: src/eth1spec/spec.py [hunk 20]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    header: `eth1spec.eth_types.Header`
+        header of interest
+    parent_header: `eth1spec.eth_types.Header`
+        parent header of header of interest
+    Returns
+    ------
+    difficulty: Uint
+        Computed difficulty for a block.
+    """
+    if parent_header.number == 0:
```
Comment (Bhargavasomu) [inline src/eth1spec/spec.py]: As per the frontier yellowpaper, shouldn't this be as follows?
```
if header.number == 0:
    # Genesis difficulty
    return Uint(131072)
```
Please correct me if wrong.
\cc @lightclient @SamWilsn

**Review by Bhargavasomu — CHANGES_REQUESTED** [review]

File: src/eth1spec/spec.py (hunk 19 above)
Comment (voith) [reply src/eth1spec/spec.py:476]: We are implementing the frontier version of the Yellow Paper. 
Screenshot 2021-06-24 at 2 19 53 PM

Comment (voith) [reply src/eth1spec/spec.py]: You are right. Thanks for catching this.

File: src/eth1spec/spec.py (hunk 20 above)
Comment (voith) [reply src/eth1spec/spec.py]: You are right

File: src/eth1spec/spec.py (hunk 7 above)
Comment (voith) [reply src/eth1spec/spec.py]: The reason why I chose this name was because the surrounding code had this name: 
https://github.com/quilt/eth1.0-specs/blob/609fdb060c6ba8101997e4e69328c840c78a135c/src/eth1spec/spec.py#L308
But I like `compute_header_hash `.

Comment (pipermerriam) [reply src/eth1spec/spec.py]: I wrote some conventions down once here: https://github.com/ethereum/snake-charmers-tactical-manual/blob/4dc11553939675e427e5daec51665033b0777538/naming-things.md#prefixes

File: tests/test_validate_header.py (hunk 18 above)
Comment (voith) [reply tests/test_validate_header.py]: I intended to write these as unit tests. IIUC, `ethereum/tests` are integration tests. 
However,  I will replace `create_fake_parent_header `  and `create_fake_header ` by parsing some fixture from `ethereum/tests`.

File: src/eth1spec/spec.py (hunk 20 above)
Comment (lightclient) [reply src/eth1spec/spec.py]: Nice catch @Bhargavasomu.

File: tests/test_validate_header.py (hunk 18 above)
Comment (lightclient) [reply tests/test_validate_header.py]: There are some unit tests, like the `TrieTests`. But you're right, they're mostly integration tests. Since this will be "the spec" though, I think we need avoid have any tests that aren't readily available to run on other clients. Ideally, any unit tests we write we should PR into `ethereum/tests` and then test them using that mechanism. If we really need some unit tests for things right now, I suppose we can leave them. But I think our longer-term goal will be to move them to `ethereum/tests`, so I don't want to spend too much time on them.

Comment (lightclient) [reply tests/test_validate_header.py]: cc: @SamWilsn if you have thoughts on this

Comment (SamWilsn) [reply tests/test_validate_header.py]: Personally I'd expect to get ~90% coverage from running `ethereum/tests`, but we might need some unit tests to get to 100%.

I don't have a problem with running Python tests, as long as the vast majority of our coverage comes from `ethereum/tests`.

File: src/eth1spec/spec.py [hunk 21]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+    valid : `bool`
+        True if gas limit constraints are satisfied, False otherwise
+    """
+    max_adjustment_delta = parent_gas_limit // GAS_LIMIT_ADJUSTMENT_FACTOR
+    if gas_limit >= parent_gas_limit + max_adjustment_delta:
+        return False
+    if gas_limit <= parent_gas_limit - max_adjustment_delta:
+        return False
+    if gas_limit < GAS_LIMIT_MINIMUM:
+        return False
```
Comment (SamWilsn) [inline src/eth1spec/spec.py]: In `verify_header`, you do a series of `assert` statements followed by a `return True`, but here you use a series of `if` which can return `False`.

Is there any reason for the different styles?

File: src/eth1spec/spec.py [hunk 22]
```diff
@@ -333,6 +348,143 @@ def signing_hash(tx: Transaction) -> Hash32:
...
+        header of interest
+    parent_header: `eth1spec.eth_types.Header`
+        parent header of header of interest
+    Returns
+    ------
+    difficulty: Uint
+        Computed difficulty for a block.
+    """
+    if header.number == 0:
+        return Uint(131072)
```
Comment (SamWilsn) [inline src/eth1spec/spec.py]: Might look ~~prettier~~ less magic as a hexadecimal number?

**Review by SamWilsn — COMMENTED** [review]
Looks pretty decent to me, just a nit and a style question.

File: src/eth1spec/spec.py (hunk 21 above)
Comment (pipermerriam) [reply src/eth1spec/spec.py]: I would suggest this as a straw-man starting point for a convention for this codebase:

```python
def check_thing(...) -> bool:
    ...

def validate_thing(...) -> None:
    assert ...

is_thing_valid = check_thing(...)
if check_thing(...):
   ...
assert check_thing(...)

validate_thing(...)  # blows out an AssertionError or passes along quietly.
```

As an implementation detail: you can compose these like this:

```python
def check_thing(...) -> bool:
    ...  # actual implementation here
    
def validate_thing(...) -> None:
    assert check_thing(...)
```

This type of composition is "convenient" but it also loses any exception context since all you get is a `True/False`.  This can be inverted as:

```python
def validate_thing(...) -> None:
    assert condition

def check_thing(...) -> bool:
    try:
        validate_thing(...)
    except AssertionError:
        return True
    else:
        return False
```

This preserves detailed error information while still avoiding the need to duplicate logic.

File: src/eth1spec/spec.py (hunk 17 above)
Comment (pipermerriam) [reply src/eth1spec/spec.py]: I think this is probably a good pattern (passing in explicit values as opposed to the full header object).  This avoids complex future typing problems where you end up trying to retain this function signature and ending up with complex types for your `Header` definition.  Keeps you from "pretending" that some of these validation functions actually have common signatures across different forks.

**Comment by voith** [comment]
Hey guys, I'm sorry for keeping this PR on hold for so long. I took the vaccine on Saturday and I've been having fever and headache. I tried to work on this today but I can't stare at the screen for too long. I'm hoping to recover by tomorrow and then address all the comments on this PR.

**Comment by SamWilsn** [comment]
> Hey guys, I'm sorry for keeping this PR on hold for so long. I took the vaccine on Saturday and I've been having fever and headache. I tried to work on this today but I can't stare at the screen for too long. I'm hoping to recover by tomorrow and then address all the comments on this PR.

Don't stress about this at all! Take care of yourself, and we'll see you when you're feeling better :purple_heart: 

You might have a couple merge conflicts though ;)

**Comment by lightclient** [comment]
+1 take care of yourself @voith! The spec will be here when you recover :)

**Review by SamWilsn — APPROVED** [review]
LGTM, @lightclient?

**Review by lightclient — APPROVED** [review]
:rocket:
